RecoLocalCalo/HcalRecProducers/src/HBHEPhase1Reconstructor.cc

```diff
@@ -121,6 +127,48 @@ namespace {
     {
         return std::pair<bool,bool>(df.linkError(), df.capidError());
     }
+
+    std::unique_ptr<HBHEStatusBitSetter> parse_HBHEStatusBitSetter(
+        const edm::ParameterSet psdigi)
+    {
+        return std::make_unique<HBHEStatusBitSetter>(
+            psdigi.getParameter<double>("nominalPedestal"),
+            psdigi.getParameter<double>("hitEnergyMinimum"),
+            psdigi.getParameter<int>("hitMultiplicityThreshold"),
+            psdigi.getParameter<std::vector<edm::ParameterSet> >("pulseShapeParameterSets"));
+    }
+
+    std::unique_ptr<HBHEPulseShapeFlagSetter> parse_HBHEPulseShapeFlagSetter(
+        const edm::ParameterSet &psPulseShape)
+    {
+        return std::make_unique<HBHEPulseShapeFlagSetter>(
+            psPulseShape.getParameter<double>("MinimumChargeThreshold"),
+            psPulseShape.getParameter<double>("TS4TS5ChargeThreshold"),
```

Comment:
I am not particularly impressed with the discussion in

http://stackoverflow.com/questions/4592352/number-of-parameters-for-a-constructor

The constructor, obviously, needs as many arguments
as the complexity of the object requires. For designs not
involving inheritance, "Builder" looks like an anti-pattern,
leading to an extra useless class. Method chaining is
not needed if the variables used in the constructor call
are appropriately named.

I do agree that, if some CMSSW object is itself constructed
only in the constructor of the framework module, it is useful
to have a constructor taking |edm::ParameterSet const&|
as a single argument.

On 09/08/2016 09:57 AM, Slava Krutelyov wrote:
>
> In RecoLocalCalo/HcalRecProducers/src/HBHEPhase1Reconstructor.cc <https://github.com/cms-sw/cmssw/pull/15779#discussion_r78023168>:
>
> > +    std::unique_ptr<HBHEStatusBitSetter> parse_HBHEStatusBitSetter(
> > +        const edm::ParameterSet psdigi)
> > +    {
> > +        return std::make_unique<HBHEStatusBitSetter>(
> > +            psdigi.getParameter<double>("nominalPedestal"),
> > +            psdigi.getParameter<double>("hitEnergyMinimum"),
> > +            psdigi.getParameter<int>("hitMultiplicityThreshold"),
> > +            psdigi.getParameter<std::vector<edm::ParameterSet> >("pulseShapeParameterSets"));
> > +    }
> > +
> > +    std::unique_ptr<HBHEPulseShapeFlagSetter> parse_HBHEPulseShapeFlagSetter(
> > +        const edm::ParameterSet &psPulseShape)
> > +    {
> > +        return std::make_unique<HBHEPulseShapeFlagSetter>(
> > +            psPulseShape.getParameter<double>("MinimumChargeThreshold"),
> > +            psPulseShape.getParameter<double>("TS4TS5ChargeThreshold"),
>
> To clarify, my original comment about reduction to the number of arguments in function calls was going towards all the code in your purview ( RecoLocalCalo/HcalRec* etc) if you get to refactoring it.
>
> Specific to these lines.
> http://stackoverflow.com/questions/4592352/number-of-parameters-for-a-constructor
> gives the same conclusion that 20-some parameters are too many.
> My solution with a struct with as many setting params is not appropriate for the constructor call.
>
> Your |parse_HBHEPulseShapeFlagSetter(edm::ParameterSet const&)| is definitely an improvement though, compared to explicit constructor calls in possibly various places.
> this can be made a part of HBHEPulseShapeFlagSetter
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/cms-sw/cmssw/pull/15779/files/d201d93f931d7a63c81f8689979d4ffd6cdedbeb#r78023168>, or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AFS4-Rhi1H6Tg3iWnfdwzgKMh7sDNBkHks5qoCJOgaJpZM4J36UI>.
>